bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/IStructuredSelection.java

```diff
@@ -15,20 +15,40 @@
 
 import java.util.Iterator;
 import java.util.List;
+import java.util.Optional;
 
 /**
  * A selection containing elements.
  */
 @SuppressWarnings("rawtypes")
 public interface IStructuredSelection extends ISelection, Iterable {
 	/**
-	 * Returns the first element in this selection, or <code>null</code>
-	 * if the selection is empty.
+	 * Returns the first element in this selection, or <code>null</code> if the
+	 * selection is empty.
 	 *
 	 * @return an element, or <code>null</code> if none
 	 */
 	public Object getFirstElement();
 
+	/**
+	 * Returns an optional describing the first element of the requested type or an
+	 * empty optional if the selection is empty or does not contain any compatible
+	 * element in regard to the specified type
+	 *
+	 * @param type the expected type
+	 * @param <T>  the desired type
+	 * @return an optional describing the result
+	 * @since 3.32
+	 */
+	default <T> Optional<T> getFirstElement(Class<T> type) {
```

Comment:
`Stream` is a separate thing for me and last time I tried this a year ago or so it resulted in endless discussions so I'd better like to start with simpler things now.

Also it *might* be more performant for some implementations to only fetch a few / items than prepare a full stream.

Yes the method has a slightly more relaxed semantics here but I assume this is never an issue anyways.

Reply:
I read `getFirstElement` as the implementation tells, otherwise, I would expect the method to be named `getFirstElementAs(Class<T>)`. Indeed most of the current code can be substituted by some `getFirstElementAs(Class<T>)` without changing the behavior, it would be safer to keep the behavior for consumers.
@laeubi Are you aware of many consumers getSelection actually iterate over the whole selection filtering for a particular type?

About going to streams, we can also do something more precise or powerful technically, but here the proposal is meant to replace the very common pattern we see everywhere. If the new API is noticeably more complex than the legacy pattern, then the API proposal is less beneficial as it will be less used...

Reply:
@sratz I renamed the method now to `getFirstElementOf(...)` to make the difference more clear.

Reply:
Most consumer simply want *an element* in contrast to really *the first element*, also most viewers only contain the same type of elements (even though elements might have different subclasses), I just wanted to avoid making it more complex than it actually is in practice ... I hope the rename makes it now more clear and easy to understand.

Reply:
>  I hope the rename makes it now more clear and easy to understand.

The issue with this method as opposed to `getFirstElementAs` is that consumers cannot just replace safely: if they're in a strange state where they care about the 1st element only (which is the one implemented in the vast majority of consumers), and the 1st is not of the right type but the 2nd is, then adopting this method will cause a change in the behavior.
So if this is not an equivalent replacement, it's much harder to adopt it safely.

Reply:
> I think it is safe in almost all cases,

While I agree, there is a major different between "we think it is safe in almost all cases" and "we're certain it's safe in all cases" ;)